clinical/src/main/java/com/hartwig/actin/clinical/curation/CurationDatabaseValidator.java

```diff
@@ -0,0 +1,119 @@
+
+    static final String GENERIC_PARENT_DOID = "4"; // "disease"
+    static final String CANCER_PARENT_DOID = "14566"; // "disease of cellular proliferation"
+
+    @NotNull
+    private final DoidModel doidModel;
+
+    public CurationDatabaseValidator(@NotNull final DoidModel doidModel) {
+        this.doidModel = doidModel;
+    }
+
+    public void validate(@NotNull CurationDatabase database) {
+        LOGGER.info("Running curation database validation");
+
+        boolean allPrimaryTumorConfigsValid = validatePrimaryTumorConfigs(database.primaryTumorConfigs(), doidModel);
+        boolean allSecondPrimaryConfigsValid = validateSecondPrimaryConfigs(database.secondPrimaryConfigs(), doidModel);
+        boolean allNonOncologicalHistoryConfigsValid =
+                validateNonOncologicalHistoryConfigs(database.nonOncologicalHistoryConfigs(), doidModel);
+        boolean allIntoleranceConfigsValid = validateIntoleranceConfigs(database.intoleranceConfigs(), doidModel);
+
+        if (allPrimaryTumorConfigsValid && allSecondPrimaryConfigsValid && allNonOncologicalHistoryConfigsValid
+                && allIntoleranceConfigsValid) {
+            LOGGER.info(" Curation database validation found no quality issues");
+        }
+    }
+
+    @VisibleForTesting
+    static boolean validatePrimaryTumorConfigs(@NotNull List<PrimaryTumorConfig> primaryTumorConfigs,
+            @NotNull DoidModel doidModel) {
+        boolean allValid = true;
+        for (PrimaryTumorConfig primaryTumorConfig : primaryTumorConfigs) {
+            if (!hasValidDoids(primaryTumorConfig.doids(), doidModel, CANCER_PARENT_DOID)) {
+                allValid = false;
+                LOGGER.warn(" Invalid primary tumor doids configured for '{}': {}", primaryTumorConfig.input(), primaryTumorConfig.doids());
+            }
+        }
+        return allValid;
+    }
+
+    @VisibleForTesting
+    static boolean validateSecondPrimaryConfigs(@NotNull List<SecondPrimaryConfig> secondPrimaryConfigs,
+            @NotNull DoidModel doidModel) {
+        boolean allValid = true;
+        for (SecondPrimaryConfig secondPrimaryConfig : secondPrimaryConfigs) {
+            if (!hasValidDoids(secondPrimaryConfig.curated().doids(), doidModel, CANCER_PARENT_DOID)) {
+                allValid = false;
+                LOGGER.warn(" Invalid second primary doids configured for '{}': {}",
+                        secondPrimaryConfig.input(),
+                        secondPrimaryConfig.curated().doids());
+            }
+        }
+        return allValid;
+    }
+
+    @VisibleForTesting
+    static boolean validateNonOncologicalHistoryConfigs(@NotNull List<NonOncologicalHistoryConfig> nonOncologicalHistoryConfigs,
+            @NotNull DoidModel doidModel) {
+        boolean allValid = true;
+        for (NonOncologicalHistoryConfig nonOncologicalHistoryConfig : nonOncologicalHistoryConfigs) {
+            Object curated = nonOncologicalHistoryConfig.curated();
```

Comment:
I agree with your comment: the fact that this returns an `Object` makes it hard to work with. It would be better if all objects that could be returned here implemented an interface or abstract class such as `Curated` so we would have some idea of what to expect. The interface could include some useful methods like `isPriorOtherCondition()` to facilitate filtering here. That would allow something like:
```
        return nonOncologicalHistoryConfigs.stream()
                .noneMatch(nonOncologicalHistoryConfig -> Optional.ofNullable(nonOncologicalHistoryConfig.curated())
                        .filter(Curated::isPriorOtherCondition)
                        .map(curated -> (PriorOtherCondition) curated)
                        .map(PriorOtherCondition::doids)
                        .map(doids -> !hasValidDoids(doids, doidModel, GENERIC_PARENT_DOID, PRIOR_OTHER_CONDITION,
                                nonOncologicalHistoryConfig.input()))
                        .orElse(false));
```
In the above, I suggest modifying `hasValidDoids` to take some additional parameters so it can handle logging.

Reply:
I have refactored NonOncologicalHistoryConfig to contain both an optional LVEF (double) and an optional PriorOtherCondition. It's not optimal yet but at least better than Object. The current datamodel is somewhat comparable to eg ECGConfig.

I feel a more correct route would be to model every config "from the perspective of what it could become" rather than "from the perspective of where it came from". E.g. we should maintain a list of inputs with mappings of everything that should be interpreted as a PriorOtherCondition. This way every config would have an input, an ignore and one optional curated "something". Currently we modeled the config to the fields where they originate from in the questionnaire.